Basic request checking

dachrillz opened the pull request:
Add basic request body checking
We now check the request body to see if content was set. For now only
works with application/json

merge #13 first

dachrillz committed:
Add basic request body checking

We now check the request body to see if content was set. For now only
works with application/json

bittrance commented on scenarios/simple/requestBodyNotDefined/http.json:
```diff
@@ -0,0 +1,23 @@
+          "postData": {
+            "mimeType": "application/json",
+            "text": "{\"hello\":\"world\"}"
+          }
```
GET requests cannot have a body. We do not at this point validate the OpenAPI spec, but expect it to be sane, so we should not have invalid examples.

dachrillz replied:
changed the scenario to use a post

bittrance commented on src/api/model.ts:
```diff
@@ -12,7 +12,10 @@ export type HTTPMETHOD =
+  | "STATUS NOT FOUND"
+  | "REQUEST BODY NOT FOUND"
+  | "CONTENT WAS NOT FOUND"
+  | "BAD MIMETYPE";
```
Reviewing this batch of PRs, I'm realizing that we probably should formulate these in a "neutral" fashion: "The spec says X but the request has Y" sort of messages. I think we should avoid assuming that either side is right.

dachrillz replied:
agreed, but I feel like we should rewrite the whole error handling once we have the basic features in place.

RIght now I'm not entirerly sure what I actually want reported back to me.

I'm thinking that we could borrow a few ideas from how compilers report errors maybe.

bittrance commented on src/api/index.ts:
```diff
@@ -6,6 +6,14 @@ const readApi = async (path: string): Promise<a.Root> => {
+      items: Object.keys(content["application/json"].schema).length,
+    });
+  }
+};
```
Note that in the special-case of JSON, `application/json` and `text/json` are synonymous (since JSON prescribes UTF8 and ignores newlines). We can thus support both.

dachrillz replied:
Added support for text/json as well. This meant i could remove a lot of if(var === "application/json")

https://github.com/XenitAB/apidelta/pull/14/commits/f4fff90e8d058e32262067d1e4b9070cc2d7b3ec

bittrance commented on src/api/index.ts:
```diff
@@ -53,6 +58,20 @@ const getParameters = (parameters?: a.Parameter[]) => {
   });
 };
 
+const getRequestBody = (
```
I'm not sure I'm completely into the terminology, but isn't this `getRequestBodyResultNode` or something?

dachrillz replied:
I seem to have named all these get functions to reflect what is retrieved from the api specification tree. 

E.g. in this case its named after https://swagger.io/docs/specification/describing-request-body/

We could change this terminology since I agree that it is a bit confusing. 

I renamed another function that had the name node just to make the naming at least consistent

bittrance commented on src/rule/index.ts:
```diff
@@ -120,6 +120,50 @@ export const removeServer = (
+  incrementHit(operationNode.requestBody.content);
+  resultOperationNode.requestBody.content = a.newContentBody();
+
+  const mimeType = request.postData?.mimeType as string;
```
This is `requestMimeType`.

dachrillz replied:
fixed

bittrance commented on src/rule/index.ts:
```diff
@@ -120,6 +120,50 @@ export const removeServer = (
+  if (
+    mimeType === "application/json" &&
+    !operationNode.requestBody.content[mimeType]
+  ) {
```
Why is this dependent on `application/json`? The MIME type should always match whether we can validate the body?

dachrillz replied:
this was dues to some type wrangling, made it somewhat better in https://github.com/XenitAB/apidelta/pull/14/commits/f4fff90e8d058e32262067d1e4b9070cc2d7b3ec

bittrance commented on src/rule/index.ts:
```diff
@@ -195,6 +240,10 @@ const verifyRequest = (
 
+  if (request.postData) {
+    matchRequestBody(operationNode, request, resultOperationNode, result);
+  }
```
This does not cover the case that the spec says the request should have a body but it does not?

dachrillz replied:
Added a test and check for this!

bittrance reviewed (commented):
Please have a look at the comments and see if you agree that there is more to do on this PR.

dachrillz committed:
change a get to post in scenario to make more http compliant

dachrillz committed:
rename getPathNode to getPath in api parsing

A debate could be had if theses names makes sense, but at least now they
are consistent inside of this file.

dachrillz committed:
rename variable to make it more clear what it does

dachrillz committed:
Report error if request body is required but missing

dachrillz committed:
allow text/json for requestBody as well

bittrance commented on src/api/api.test.ts:
```diff
@@ -23,19 +23,11 @@ const expected: Root = {
-              },
+              "application/json": expect.objectContaining({
+                x_x_x_x_results: {
+                  hits: 0,
```
I should perhaps have caught this before, but shouldn't it start at 1? That may be a separate PR tho.

bittrance commented on src/api/model.ts:
```diff
@@ -122,15 +126,44 @@ export type Parameter = {
 export type Content = {
-  "application/json"?: any;
+  "application/json"?: MimeType;
+  "text/json"?: MimeType;
```
This looks a bit strange given that we know that these two will be identical.

bittrance reviewed (approved):
I think "start at zero" comment is still relevant, but other than that, this looks fine.
